homeassistant/components/tplink/__init__.py

```diff
@@ -126,3 +181,29 @@ def legacy_device_id(device: SmartDevice) -> str:
     if "_" not in device_id:
         return device_id
     return device_id.split("_")[1]
+
+
+def _get_store(hass: HomeAssistant) -> Store[dict[str, dict[str, str]]]:
```

Comment:
It would be better to store it in the config entry data instead so we don't have to worry about cleaning this up when the integration is removed

Reply:
For tplink there is a config_entry per device and all the authenticating devices use the same credentials.  If we store it with the config_entry data we have to update all the config_entries every time one of them changes (I actually did this in a previous PR and @rytilahti felt it was not good.  On reflection I ageed with him, you can see the discussion here https://github.com/home-assistant/core/pull/100356) and the syncing implementation in [update_listener](https://github.com/home-assistant/core/pull/100356/files#diff-d5be15ba325ec97f911ca53d9b85c32a041694ad617f75f7af52cab4e08fa1ee).

Another advantage of not putting it in the config entry data is that discovery picks up the credentials from the store and the user just needs to click confirm for new devices.

Reply:
Ok will do.  I'll also try add something to delete the store when the last config_entry is removed.

Did your KP125M arrive? If it did would you mind a quick fixture PR when you have a chance?

`python3 -m devtools.dump_devinfo YOUR_IP --username USERNAME --password PASSWORD`

Reply:
Package was delayed.  Finally showed up.  Will test in a bit when I finish up some calls

Reply:
@sdb9696 

```
Testing Call(module='component_nego', method='component_nego')..FAIL Unable to connect to the device: 192.168.107.103: 'NoneType' object has no attribute 'encode'
Testing Call(module='device_info', method='get_device_info')..FAIL Unable to connect to the device: 192.168.107.103: 'NoneType' object has no attribute 'encode'
Testing Call(module='device_usage', method='get_device_usage')..FAIL Unable to connect to the device: 192.168.107.103: 'NoneType' object has no attribute 'encode'
Testing Call(module='device_time', method='get_device_time')..FAIL Unable to connect to the device: 192.168.107.103: 'NoneType' object has no attribute 'encode'
Testing Call(module='energy_usage', method='get_energy_usage')..FAIL Unable to connect to the device: 192.168.107.103: 'NoneType' object has no attribute 'encode'
Testing Call(module='current_power', method='get_current_power')..FAIL Unable to connect to the device: 192.168.107.103: 'NoneType' object has no attribute 'encode'
Testing Call(module='temp_humidity_records', method='get_temp_humidity_records')..FAIL Unable to connect to the device: 192.168.107.103: 'NoneType' object has no attribute 'encode'
Testing Call(module='child_device_list', method='get_child_device_list')..FAIL Unable to connect to the device: 192.168.107.103: 'NoneType' object has no attribute 'encode'
Testing Call(module='trigger_logs', method={'get_trigger_logs': {'page_size': 5, 'start_id': 0}})..FAIL Unable to connect to the device: 192.168.107.103: 'NoneType' object has no attribute 'encode'
Testing Call(module='child_device_component_list', method='get_child_device_component_list')..FAIL Unable to connect to the device: 192.168.107.103: 'NoneType' object has no attribute 'encode'
Unable to query all successes at once: Unable to connect to the device: 192.168.107.103: 'NoneType' object has no attribute 'encode'
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/Users/bdraco/python-kasa/devtools/dump_devinfo.py", line 278, in <module>
    cli()
  File "/opt/homebrew/lib/python3.12/site-packages/asyncclick/core.py", line 1157, in __call__
    return anyio.run(self._main, main, args, kwargs, **opts)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.12/site-packages/anyio/_core/_eventloop.py", line 66, in run
    return async_backend.run(func, args, {}, backend_options)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.12/site-packages/anyio/_backends/_asyncio.py", line 1960, in run
    return native_run(wrapper(), debug=debug)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.12/3.12.0/Frameworks/Python.framework/Versions/3.12/lib/python3.12/asyncio/runners.py", line 194, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.12/3.12.0/Frameworks/Python.framework/Versions/3.12/lib/python3.12/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.12/3.12.0/Frameworks/Python.framework/Versions/3.12/lib/python3.12/asyncio/base_events.py", line 664, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.12/site-packages/anyio/_backends/_asyncio.py", line 1953, in wrapper
    return await func(*args)
           ^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.12/site-packages/asyncclick/core.py", line 1160, in _main
    return await main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.12/site-packages/asyncclick/core.py", line 1076, in main
    rv = await self.invoke(ctx)
         ^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.12/site-packages/asyncclick/core.py", line 1434, in invoke
    return await ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.12/site-packages/asyncclick/core.py", line 780, in invoke
    rv = await rv
         ^^^^^^^^
  File "/Users/bdraco/python-kasa/devtools/dump_devinfo.py", line 110, in cli
    save_filename, copy_folder, final = await get_smart_fixture(device)
                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/bdraco/python-kasa/devtools/dump_devinfo.py", line 254, in get_smart_fixture
    for response in responses["responses"]:
                    ^^^^^^^^^
UnboundLocalError: cannot access local variable 'responses' where it is not associated with a value
```

With the EP25

Reply:
P125M
```
 % python3 -m devtools.dump_devinfo 192.168.107.157 --username 'SNIP' --password 'SNIP'
Testing Call(module='component_nego', method='component_nego')..OK
Testing Call(module='device_info', method='get_device_info')..OK
Testing Call(module='device_usage', method='get_device_usage')..OK
Testing Call(module='device_time', method='get_device_time')..OK
Testing Call(module='energy_usage', method='get_energy_usage')..FAIL not suported
Testing Call(module='current_power', method='get_current_power')..FAIL not suported
Testing Call(module='temp_humidity_records', method='get_temp_humidity_records')..FAIL not suported
Testing Call(module='child_device_list', method='get_child_device_list')..FAIL not suported
Testing Call(module='trigger_logs', method={'get_trigger_logs': {'page_size': 5, 'start_id': 0}})..FAIL not suported
Testing Call(module='child_device_component_list', method='get_child_device_component_list')..FAIL not suported
Got 4 successes
## device info file ##
{'component_nego': {'component_list': [{'id': 'device', 'ver_code': 2},
                                       {'id': 'firmware', 'ver_code': 2},
                                       {'id': 'quick_setup', 'ver_code': 3},
                                       {'id': 'time', 'ver_code': 1},
                                       {'id': 'wireless', 'ver_code': 1},
                                       {'id': 'schedule', 'ver_code': 2},
                                       {'id': 'countdown', 'ver_code': 2},
                                       {'id': 'antitheft', 'ver_code': 1},
                                       {'id': 'account', 'ver_code': 1},
                                       {'id': 'synchronize', 'ver_code': 1},
                                       {'id': 'sunrise_sunset', 'ver_code': 1},
                                       {'id': 'led', 'ver_code': 1},
                                       {'id': 'cloud_connect', 'ver_code': 1},
                                       {'id': 'iot_cloud', 'ver_code': 1},
                                       {'id': 'device_local_time',
                                        'ver_code': 1},
                                       {'id': 'default_states', 'ver_code': 1},
                                       {'id': 'auto_off', 'ver_code': 2},
                                       {'id': 'localSmart', 'ver_code': 1},
                                       {'id': 'matter', 'ver_code': 2}]},
 'discovery_result': {'device_id': '00000000000000000000000000000000',
                      'device_model': 'P125M(US)',
                      'device_type': 'SMART.TAPOPLUG',
                      'factory_default': False,
                      'ip': '127.0.0.123',
                      'is_support_iot_cloud': True,
                      'mac': '00-00-00-00-00-00',
                      'mgt_encrypt_schm': {'encrypt_type': 'KLAP',
                                           'http_port': 80,
                                           'is_support_https': False,
                                           'lv': 2},
                      'obd_src': 'tplink',
                      'owner': '00000000000000000000000000000000'},
 'get_device_info': {'auto_off_remain_time': 0,
                     'auto_off_status': 'off',
                     'avatar': 'plug',
                     'default_states': {'state': {}, 'type': 'last_states'},
                     'device_id': '0000000000000000000000000000000000000000',
                     'device_on': True,
                     'fw_id': '00000000000000000000000000000000',
                     'fw_ver': '1.1.0 Build 231009 Rel.155831',
                     'has_set_location_info': False,
                     'hw_id': '00000000000000000000000000000000',
                     'hw_ver': '1.0',
                     'ip': '127.0.0.123',
                     'lang': 'en_US',
                     'latitude': 0,
                     'longitude': 0,
                     'mac': '00-00-00-00-00-00',
                     'model': 'P125M',
                     'nickname': 'I01BU0tFRF9OQU1FIw==',
                     'oem_id': '00000000000000000000000000000000',
                     'on_time': 41,
                     'overheated': False,
                     'region': 'Pacific/Honolulu',
                     'rssi': -51,
                     'signal_level': 2,
                     'specs': '',
                     'ssid': 'I01BU0tFRF9TU0lEIw==',
                     'time_diff': -600,
                     'type': 'SMART.TAPOPLUG'},
 'get_device_time': {'region': 'Pacific/Honolulu',
                     'time_diff': -600,
                     'timestamp': 1702079637},
 'get_device_usage': {'time_usage': {'past30': 1, 'past7': 1, 'today': 1}}}
Do you want to save the above content to P125M(US)_1.0_1.1.0.json (y/n): y
Saving info to P125M(US)_1.0_1.1.0.json
Saved. Copy/Move P125M(US)_1.0_1.1.0.json to kasa/tests/fixtures/smart/ to add it to the test suite
```

Reply:
Hi @bdraco, if you have a moment could you please try regenerating the fixtures for your EP25 and KP125M as we've added a lot more queries and I think the failure on your EP25 should hopefully be resolved.

Reply:
<img width="713" alt="Screenshot 2024-01-04 at 12 12 50 PM" src="https://github.com/home-assistant/core/assets/663432/601ceac9-4798-433b-857c-eaee4d635f83">

EP25 looks good

Reply:
https://github.com/python-kasa/python-kasa/pull/621